### DISCUSSION TITLE
can i use repository pattern on rakit?

### DISCUSSION THREAD

**Discussion by wznuhidayat** [esyede/rakit#37] [Diskusi Umum] [answered]
how i implement repository and interface on this framework?

because every i created files in my custom folder always considered as static class

**Comment by esyede** [comment]
Hi, if you need to load custom classes outside the `application/libraries` folder, you need to manually autoload it in the `application/boot.php` file just like when you create a new package. For example, if you have your repository classes in `application/repositories`

`boot.php`
```php
Autoloader::directories([
    path('app') . 'repositories',
	// other paths..
]);
```

Docs about autoloading can be found here: https://rakit.esyede.my.id/docs/id/autoloading

**Reply by wznuhidayat** [reply]
> Hi, if you need to load custom classes outside the `application/libraries` folder, you need to manually autoload it in the `application/boot.php` file just like when you create a new package. For example, if you have your repository classes in `application/repositories`
> 
> `boot.php`
> 
> ```
> Autoloader::directories([
>     path('app') . 'repositories',
> 	// other paths..
> ]);
> ```
> 
> Docs about autoloading can be found here: https://rakit.esyede.my.id/docs/id/autoloading

why always need static class?
![image](https://github.com/user-attachments/assets/bd278de2-bb2d-4c44-b572-bfb942d477e0)

![image](https://github.com/user-attachments/assets/c05ebd11-acea-4a65-817d-27ea3e978a5a)

**Reply by esyede** [reply]
Hi, you should instantiate that class first, as `getPostByID` is indeed not a static method.
```php
$testRepo = new TestRepository();
dd($testRepo->getPostByID(1));
```

You can also register the `TestRepository` class into Rakit's container if you don't want to do the instantiation multiple times,I mean, to make it a singleton:
```php
Container::singleton('testRepo', function () {
    return new TestRepository();
});

// And to use it somewhere:
$testRepo = Container::resolve('testRepo');
dd($testRepo->getPostByID(1));
```

**Reply by wznuhidayat** [reply]
ok great work for me.. 
what if i want to add some subfolder like this?
![image](https://github.com/user-attachments/assets/f4571336-97d4-4465-80b8-5e776a300445)

**Reply by esyede** [reply, marked as answer]
Hi! You can load the classes via the Autoloader. (just like when you ask about loading files inside a custom package last time). Then you can use your classes as usual.

Docs for autoloder can be found here: 
https://rakit.esyede.my.id/docs/id/autoloading
